### PR TITLE
Add support for Certificate management in Administration/Certificate Library 

### PR DESCRIPTION
**Pull request by vbauzysvmware**
Added types:
* `Certificate` and `types.CertificateLibraryItem` for handling Certificates in Certificate Library 

Functions:
* `GetCertificateFromLibraryById`, `AddCertificateToLibrary`, `GetAllCertificatesFromLibrary`, `GetCertificateFromLibraryByName`, `Update`, `Delete`

### CODE REVIEW

File: govcd/openapi.go [hunk 1]
```diff
@@ -628,7 +628,14 @@ func (client *Client) newOpenApiRequest(apiVersion string, params url.Values, me
...
 
 	// Add the params to our URL
-	reqUrlCopy.RawQuery += params.Encode()
+
+	// url.QueryEscape as well as url.Values.Encode()
+	// both encode the space as a + character
+	// https://github.com/golang/go/issues/4013
+	// https://github.com/czos/goamz/pull/11/files
+	// additional strings.Replace change encoded %2B(+) to %20(space)
+
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/openapi.go:637]: removed

File: govcd/openapi.go [hunk 2]
```diff
@@ -628,7 +628,14 @@ func (client *Client) newOpenApiRequest(apiVersion string, params url.Values, me
 	reqUrlCopy := copyUrlRef(reqUrl)
 
 	// Add the params to our URL
-	reqUrlCopy.RawQuery += params.Encode()
+
+	// url.QueryEscape as well as url.Values.Encode()
+	// both encode the space as a + character
+	// https://github.com/golang/go/issues/4013
+	// https://github.com/czos/goamz/pull/11/files
+	// additional strings.Replace change encoded %2B(+) to %20(space)
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/openapi.go:636]: changed

File: govcd/openapi.go [hunk 3]
```diff
@@ -628,7 +628,14 @@ func (client *Client) newOpenApiRequest(apiVersion string, params url.Values, me
 	reqUrlCopy := copyUrlRef(reqUrl)
 
 	// Add the params to our URL
-	reqUrlCopy.RawQuery += params.Encode()
+
+	// url.QueryEscape as well as url.Values.Encode()
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/openapi.go:632]: changed

File: govcd/openapi.go [hunk 4]
```diff
@@ -628,7 +628,14 @@ func (client *Client) newOpenApiRequest(apiVersion string, params url.Values, me
...
+	// both encode the space as a + character
+	// https://github.com/golang/go/issues/4013
+	// https://github.com/czos/goamz/pull/11/files
+	// additional strings.Replace change encoded %2B(+) to %20(space)
+
+	reqUrlCopy.RawQuery += strings.Replace(params.Encode(), "%2B", "%20", -1)
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/openapi.go:636]: well, names with `+` won't work. Standard library functions change space to +, so later we don't know if + is real of was space.

[the remark this answers is not in the document]
Comment (dataclouder) [reply govcd/openapi.go:636]: I see that the [referred issue](https://github.com/golang/go/issues/4013) says that representing a space with `+` in query parameters is working as expected.
I would recommend leaving the encoding as it was, and  fix any problems with spaces if they arise.
In both cases (either using the manual replacement or reverting to the previous one) we should have specific tests to make sure the Open API engine behaves as expected with the queries that we are submitting

[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/openapi.go]: As we agreed on chat, I moved the solution to the certificate fetch function, due to that there is a bug in the upstream library and we are not sure the generic solution is a good way.

**Review by lvirbalas — APPROVED** [review]

**Pull request merged**